File: python-design-patterns/creational_patterns/abstract_factory/abstract_factory.py
from abc import ABC, abstractmethod


class AbstractFactory(ABC):
    @abstractmethod
    def create_product_a(self):
        pass

    @abstractmethod
    def create_product_b(self):
        pass


class ConcreteFactory1(AbstractFactory):
    def create_product_a(self):
        return ProductA1()

    def create_product_b(self):
        return ProductB1()


class ConcreteFactory2(AbstractFactory):
    def create_product_a(self):
        return ProductA2()

    def create_product_b(self):
        return ProductB2()


class ProductA(ABC):
    @abstractmethod
    def do_something(self):
        pass


class ProductA1(ProductA):
    def do_something(self):
        print("I am ProductA1")


class ProductA2(ProductA):
    def do_something(self):
        print("I am ProductA2")


class ProductB(ABC):
    @abstractmethod
    def do_something(self):
        pass


class ProductB1(ProductB):
    def do_something(self):
        print("I am ProductB1")


class ProductB2(ProductB):
    def do_something(self):
        print("I am ProductB2")


def main():
    factory = ConcreteFactory1()
    product_a = factory.create_product_a()
    product_b = factory.create_product_b()

    product_a.do_something()
    product_b.do_something()

    factory = ConcreteFactory2()
    product_a = factory.create_product_a()
    product_b = factory.create_product_b()

    product_a.do_something()
    product_b.do_something()


if __name__ == "__main__":
    main()

File: scala-design-patterns/src/main/scala/CreationalPatterns/AbstractFactory/AbstractFactory.scala
abstract class AbstractFactory {
  def createProductA(): ProductA
  def createProductB(): ProductB
}

class ConcreteFactory1 extends AbstractFactory {
  def createProductA(): ProductA = new ProductA1()
  def createProductB(): ProductB = new ProductB1()
}

class ConcreteFactory2 extends AbstractFactory {
  def createProductA(): ProductA = new ProductA2()
  def createProductB(): ProductB = new ProductB2()
}

trait ProductA {
  def doSomething(): Unit
}

class ProductA1 extends ProductA {
  def doSomething(): Unit = {
    println("I am ProductA1")
  }
}

class ProductA2 extends ProductA {
  def doSomething(): Unit = {
    println("I am ProductA2")
  }
}

trait ProductB {
  def doSomething(): Unit
}

class ProductB1 extends ProductB {
  def doSomething(): Unit = {
    println("I am ProductB1")
  }
}

class ProductB2 extends ProductB {
  def doSomething(): Unit = {
    println("I am ProductB2")
  }
}

object AbstractFactoryPattern {
  def main(args: Array[String]): Unit = {
    val factory1 = new ConcreteFactory1()
    val productA1 = factory1.createProductA()
    val productB1 = factory1.createProductB()

    productA1.doSomething()
    productB1.doSomething()

    val factory2 = new ConcreteFactory2()
    val productA2 = factory2.createProductA()
    val productB2 = factory2.createProductB()

    productA2.doSomething()
    productB2.doSomething()
  }
}